### ISSUE TITLE
Failed to locate opened application window with appId on Remote Desktop

### ISSUE THREAD

**Issue by tienledrake** [microsoft/WinAppDriver#1748]
My test runs successfully on my local machine but it gets "Failed to locate opened application window with appId" error while running on Remote Desktop through Release Pipeline on Azure DevOps. 

Here is my InitialiizeApp() method:
        public static void InitializeApp()
        {
            var driverOptions = new AppiumOptions();
            driverOptions.AddAdditionalCapability("deviceName", "WindowsPC");
            driverOptions.AddAdditionalCapability("platformName", "Windows");
            driverOptions.AddAdditionalCapability("appWorkingDir", Settings.AppFolder);
            driverOptions.AddAdditionalCapability("app", Settings.App);

            DriverContext.Driver = new WindowsDriver<WindowsElement>(new Uri(Settings.WindowsApplicationDriverUrl), driverOptions);
            DriverContext.Driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
        }

Does anyone face the same issue? Please help.

**Comment by kamilchodola** [comment]
Hi!

This happens because Your application is not opened at all or time spent on waiting for app to open is too long.
I had exactly the same issue on my solution.
There are two options to solve that:

1. Add additional app capability: 
`appCapabilities.SetCapability("ms:waitForAppLaunch", "25");`
This will add additional waiting time before driver will throw this issue.
There is one big issue of that solution: this timeout is not dynamic but static - even if application will open properly after 10 seconds, it will wait until 25 second time elapse.
2. Other (a bit dirty) way around which I've applied on my solution is to check if process is started and have MainWindowHandle. If yes, then initialize the driver again, but using appTopLevelWIndow capability:
        
        public static WindowsDriver<WindowsElement> InitializeDriver(Uri uri, AppiumOptions options)
        {
            WindowsDriver<WindowsElement> driver = null;

            try
            {
                driver = new WindowsDriver<WindowsElement>(uri, options);
            }
            catch (WebDriverException e)
            {
                if (e.Message.Contains("Failed to locate opened application"))
                {
                    int retryCount = 0;
                    string value = options.ToCapabilities().GetCapability("app").ToString().Split('/').ToList().Last().Replace(".exe", "");
                    var processes = Process.GetProcessesByName(value);
                    while (processes[0].MainWindowHandle.ToString("x") == "0" && retryCount < 30)
                    {
                        retryCount++;
                        Thread.Sleep(1000);
                        processes = Process.GetProcessesByName(value);
                    }
                    if (processes[0].MainWindowHandle.ToString("x") == "0")
                        throw e;

                    var newOptions = new AppiumOptions();
                    newOptions.AddAdditionalCapability("appTopLevelWindow", processes[0].MainWindowHandle.ToString("x"));
                    driver = new WindowsDriver<WindowsElement>(uri, newOptions);
                }
                else
                {
                    throw e;
                }
            }
            return driver;
        }
Maybe it is not the most elegant way but most important thing is that this is working properly and is dynamic (retryCount may be switched for some better way of time pooling management).

**Comment by tienledrake** [comment]
Thank kamilchodola for your response.
My application is not opened at all. The weird thing is that the issue happens only on remote desktop while I run release pipeline on Azure DevOps, but not on my local machine.

I tried both options you suggested but the same issue still happens on remote desktop. Below is the detail error.

![image](https://user-images.githubusercontent.com/103521457/176446824-79e94981-354a-44e0-903c-58e132c3e102.png)

Thank you so much for your help.

**Comment by liljohnak** [comment]
@tienledrake we use similar code to above to make it work.  When your application starts, does it have multiple processIds (maybe 1 without a window) ?  The above error suggests that the process that you are testing does not have a Window within 10 seconds.

**Comment by kamilchodola** [comment]
Yeah, I suggest checking the case described by @liljohnak - In my app we had two windows opened from one shortcut and it happened that wrong window was chosen (console was opened at the beggining). Solution was to add runparameter for application which prevent from opening the console window.

Also - if it works on local and does not on remote - check paths. Those most probably are different on remote machine.
One more thing regarding app and appWorkingDir - both needs to use absolute path rather than relative (for app use C:\Windows\calc.exe insetd of calc.exe). We had some issues with running app when relative was used even when working dir was set.

Is the application You are testing have a long launch time on local machine?

**Comment by tienledrake** [comment]
@kamilchodola I tried the second option you gave me, this time I got different error:

_System.IndexOutOfRangeException : Index was outside the bounds of the array.

at the "while (processes[0].MainWindowHandle.ToString("x") == "0" && retryCount < 30)" row.

**Comment by kamilchodola** [comment]
@tienledrake 
Yeah - missing part here is verification if processes are not empty - but this give a nice hint here - your process is not starting at all.
Still seems like a issue with a startup parameters.
        driverOptions.AddAdditionalCapability("appWorkingDir", Settings.AppFolder);
        driverOptions.AddAdditionalCapability("app", Settings.App);

Are those two Settings are somehow overriden while running tests at remote machine? Variables in pipeline or You have some app.settings in Your solution?
Maybe You can add some Console.WriteLine for both of those params and push such version to run on remote using pipeline and see if those values are for sure correct.

**Comment by kamilchodola** [comment]
@tienledrake 
Yeah it is.
So the things which came to my mind about that are:
1. Check in Event history right after text execution if there is any error about running the app (maybe some priviliges issue for agent or something simmilar).
2. Try running tests on VM but using console runner to see if the error will be reproduced.
3. You can try to make a different approach - Run application using Process.Start() from c# code and then attach to it using this snippet:
    `var processes = Process.GetProcessesByName(value);`
    `newOptions.AddAdditionalCapability("appTopLevelWindow", processes[0].MainWindowHandle.ToString("x"));`
    `driver = new WindowsDriver<WindowsElement>(uri, newOptions);`
You can add to it additional verification if processes are not empty List and if so, wait for a couple of seconds (even using hardcoded Thread.Sleep for test purpose).
4. The last thing I have on my mind is to Install VS on VM and check if tests would work using builtin runner

**Comment by tienledrake** [comment]
Thanks a lot for your suggestion, @kamilchodola 

I followed your instruction. There is an error in Event History that makes the application failed to start. So it does not relate to the automation script, it is because of some dependencies need to be installed to start .Net application. Error as below:

 _Failed to load the dll from [C:\agent\_work\_tool\dotnet\host\fxr\6.0.6\hostfxr.dll]_

**Comment by tienledrake** [comment]
@kamilchodola VM has .NET 6, if I double-click on the .exe file manually, the application open normally. But when using Release Pipeline, it got that issue even though I have the step to install .NET SDK 6.0.301 in the pipeline. It's weird.

**Comment by kamilchodola** [comment]
@tienledrake 
So maybe try opposite way - remove this step (do not force any dotnet version - let VM choose) and see how it would work.
Checking on the internet there is a lot of issues with this dll file on various version - maybe latest one have some problems with it as well.

Another thing - You can check if there is this dll mentioned in this directory - maybe there is a bit different path and maybe You need to temporarily use different build of .NET 6 - not sure here.

**Comment by tienledrake** [comment]
@kamilchodola You saved my life. It works as expected now, after I disabled the step to install .NET 6 for agent in Release pipeline.

Thank you so much. I really appreciate your help, @kamilchodola

**Comment by gabrieldelaparra** [comment]
This worked for me:
```
options.AddAdditionalCapability("ms:waitForAppLaunch", "3");
```
